main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -384,20 +382,83 @@ object Safety {
       val args2 = tpe2.arrowArgTypes
       val superTypeArgs = args1.zip(args2).forall {
         case (t1, t2) =>
-          isSubTypeOf(t2, t1, renv)
+          isSubtypeOf(t2, t1, renv)
       }
 
       // check that result is a subtype
       val expectedResTpe = tpe1.arrowResultType
       val actualResTpe = tpe2.arrowResultType
-      val subTypeResult = isSubTypeOf(expectedResTpe, actualResTpe, renv)
+      val subTypeResult = isSubtypeOf(expectedResTpe, actualResTpe, renv)
 
       subTypePurity && isEffSubset && superTypeArgs && subTypeResult
 
     case _ => tpe1 == tpe2
 
   }
 
+  /**
+    * ADT to indicate what the result of checking a java subtype was.
+    */
+  private sealed trait JavaSubtypeResult
+
+  private case object Castable extends JavaSubtypeResult
+
+  private case object NonCastable extends JavaSubtypeResult
+
+  /**
+    * @param tpe the non java type.
+    */
+  private case class NonJavaType(tpe: Type) extends JavaSubtypeResult
+
+  private case object TypeVariable extends JavaSubtypeResult
+
+  /**
+    * Returns true if `tpe1` and `tpe2` are both java types
+    * and `tpe1` is a subtype of `tpe2`.
+    */
+  private def isJavaSubtypeOf(tpe1: Type, tpe2: Type)(implicit flix: Flix): JavaSubtypeResult = (tpe1.baseType, tpe2.baseType) match {
+    case (Type.Cst(TypeConstructor.Native(left), _), Type.Cst(TypeConstructor.Native(right), _)) =>
+      if (right.isAssignableFrom(left)) Castable else NonCastable
+
+    case (_, Type.Var(_, _)) =>
+      TypeVariable
+
+    case (Type.Cst(TypeConstructor.Native(_), _), _) =>
+      NonJavaType(tpe2)
+
+    case (_, Type.Cst(TypeConstructor.Native(_), _)) =>
+      NonJavaType(tpe1)
+
+    case _ => NonCastable
+  }
+
+  /**
+    * Returns a list of errors if the the upcast is invalid.
+    */
+  private def checkUpcastSafety(exp: Expression, tpe: Type, renv: RigidityEnv, loc: SourceLocation)(implicit flix: Flix): List[SafetyError] = {
+    val tpe1 = Type.eraseAliases(exp.tpe)
+    val tpe2 = Type.eraseAliases(tpe)
+    if (isSubtypeOf(tpe1, tpe2, renv))
+      Nil
+    else
+      UnsafeUpcast(exp.tpe, tpe, loc) :: Nil
+  }
+
+  /**
+    * Returns a list of errors if the the supercast is invalid.
+    */
+  private def checkSupercastSafety(exp: Expression, tpe: Type, loc: SourceLocation)(implicit flix: Flix): List[SafetyError] = {
+    val tpe1 = Type.eraseAliases(exp.tpe)
+    val tpe2 = Type.eraseAliases(tpe)
+    isJavaSubtypeOf(tpe1, tpe2) match {
+      case Castable => Nil
+      case NonCastable => UnsafeSupercast(exp.tpe, tpe, loc) :: Nil
+      case NonJavaType(t) if t == tpe1 => NonJavaTypeSupercast(exp.tpe, tpe, loc) :: Nil
+      case NonJavaType(_) => NonJavaTypeSupercast(tpe, exp.tpe, loc) :: Nil
```

Comment:
The `t == tpe1` check seems strange to me. Also, it looks like this will be backwards half the time, no?

Do we need a specific "non-java-type" error? If we do, maybe we should just split it in two: cast to non-java-type and cast from non-java-type.